tools/flow_grid.py

```diff
@@ -0,0 +1,689 @@
+                             i - 1 + 0,
+                             i - 1 - padshape[1]]).T
+
+        # for each cell j, recursively search through grid to determine
+        # if surrounding cells are in the contributing area, then add
+        # flattened indices to self.collect
+        def catchment_search(j):
+            self.collect = np.append(self.collect, j)
+            selection = select_surround_ravel(j)
+            next_idx = selection[np.where(self.cdir[selection] == dirmap)]
+            if next_idx.any():
+                return catchment_search(next_idx)
+
+        # call catchment search starting at the pour point
+        catchment_search(pour_point)
+
+        # initialize output array
+        outcatch = np.zeros(padshape, dtype=int)
+
+        # if nodata is not 0, replace 0 with nodata value in output array
+        if nodata != 0:
+            np.place(outcatch, outcatch == 0, nodata)
+
+        # set values of output array based on 'collected' cells
+        outcatch.flat[self.collect] = self.cdir[self.collect]
+
+        # remove outer rim, delete temporary arrays
+        outcatch = outcatch[1:-1, 1:-1]
+        del self.cdir
+        del self.collect
+
+        # if pour point needs to be a special value, set it
+        if pour_value is not None:
+            outcatch[y,x] = pour_value 
+
+        # if inplace is True, update attributes
+        if inplace == True:
+            self.catch = outcatch
+            self.nodata.update({'catch' : nodata})
+            self.gridlist.append('catch')
+        else:
+            return outcatch
+
+    def fraction(self, other, nodata=0, inplace=True):
+        """
+        Generates a grid representing the fractional contributing area for a
+        coarse-scale flow direction grid.
+
+        Parameters
+        ----------
+        other : flow_grid instance
+                Another flow_grid instance containing fine-scale flow direction
+                data. The ratio of self.cellsize/other.cellsize must be a
+                positive integer. Grid cell boundaries must have some overlap.
+                Must have attributes 'dir' and 'catch' (i.e. must have a flow
+                direction grid, along with a delineated catchment).
+                
+        inplace : bool (optional)
+                  If True, appends fraction grid to attribute 'frac'.
+        """
+
+        # check for required attributes in self and other
+        assert hasattr(self, 'dir')
+        assert hasattr(other, 'dir')
+        assert hasattr(other, 'catch')
+
+        # set scale ratio
+        cell_ratio = int(self.cellsize/other.cellsize)
+
+        # create DataFrames for self and other with geographic coordinates
+        # as row and column labels. entries in selfdf represent cell indices.
+        selfdf = pd.DataFrame(
+                np.arange(self.dir.size).reshape(self.shape),
+                index=np.linspace(self.bbox[1], self.bbox[3],
```

Comment:
Indentation here is off.  Should either align with parenthesis or be one tab (4 spaces) from `selfdf`.  Same for lines 463:468.